// @darrellvs/node-wave-link-sdk/dist/index.mjs
// src/Base/BaseWaveLinkController.ts
import { WebSocket } from "ws";
import { readFileSync, existsSync } from "fs";
import { join } from "path";

// src/Helpers/EventEmitterHelpers.ts
import { EventEmitter } from "events";
var TypedEventEmitter = class {
  eventEmitter = new EventEmitter();
  constructor() {
  }
  emit(event, ...args) {
    this.eventEmitter.emit(event, ...args);
  }
  on(event, listener) {
    this.eventEmitter.on(event, listener);
  }
  off(event, listener) {
    this.eventEmitter.off(event, listener);
  }
  removeAllListeners(event) {
    this.eventEmitter.removeAllListeners(event);
  }
  setMaxListeners(n) {
    this.eventEmitter.setMaxListeners(n);
  }
};

// src/Base/BaseWaveLinkController.ts
var BaseWaveLinkController = class _BaseWaveLinkController extends TypedEventEmitter {
  ws;
  lastId = 0;
  pending = /* @__PURE__ */ new Map();
  notifications = /* @__PURE__ */ new Map();
  isConnecting = false;
  currentPort = 0;
  intentionalDisconnect = false;
  // --- Port discovery ---
  static getWsInfoPath() {
    if (process.platform === "win32") {
      const localAppData = process.env.LOCALAPPDATA ?? (process.env.APPDATA ? join(process.env.APPDATA, "../Local") : null);
      if (localAppData) {
        return join(
          localAppData,
          "Packages/Elgato.WaveLink_g54w8ztgkx496/LocalState/ws-info.json"
        );
      }
    }
    return null;
  }
  static readPortFromFile() {
    try {
      const p = _BaseWaveLinkController.getWsInfoPath();
      if (!p || !existsSync(p))
        return null;
      const data = JSON.parse(readFileSync(p, "utf-8"));
      if (typeof data.port === "number" && data.port > 0)
        return data.port;
    } catch {
    }
    return null;
  }
  // --- Connection ---
  connect() {
    if (this.isConnecting)
      return;
    this.isConnecting = true;
    const portFromFile = _BaseWaveLinkController.readPortFromFile();
    if (portFromFile) {
      this.currentPort = portFromFile;
    } else {
      this.currentPort = 1884;
    }
    this.tryConnect();
  }
  disconnect() {
    this.intentionalDisconnect = true;
    this.ws?.close();
  }
  tryConnect() {
    const url = `ws://127.0.0.1:${this.currentPort}`;
    const ws = new WebSocket(url, { headers: { Origin: "streamdeck://" } });
    this.ws = ws;
    ws.onopen = () => {
      this.isConnecting = false;
      this.onConnect();
    };
    ws.onmessage = (evt) => {
      this.resolveMessage(evt.data.toString());
    };
    ws.onerror = () => {
    };
    ws.onclose = () => {
      if (this.ws !== ws)
        return;
      this.ws = void 0;
      this.pending.forEach((p) => p.reject(new Error("WebSocket closed")));
      this.pending.clear();
      this.onDisconnect();
      this.emit("close");
      if (!this.intentionalDisconnect) {
        setTimeout(() => {
          this.isConnecting = false;
          this.connect();
        }, 1e3);
      }
      this.intentionalDisconnect = false;
    };
  }
  nextPort() {
    this.currentPort = this.currentPort >= 1893 ? 1884 : this.currentPort + 1;
  }
  // Called by subclass to skip this port and try the next one.
  // Used when a port responds but getApplicationInfo returns a non-EWL appID.
  ignorePortAndReconnect() {
    this.ws?.close();
    this.nextPort();
    this.tryConnect();
  }
  // --- JSON-RPC send ---
  send(method, params = {}) {
    return new Promise((resolve, reject) => {
      if (!this.ws || this.ws.readyState !== WebSocket.OPEN) {
        reject(new Error("Not connected"));
        return;
      }
      const id = ++this.lastId;
      this.pending.set(id, {
        resolve,
        reject
      });
      this.ws.send(JSON.stringify({ jsonrpc: "2.0", method, params, id }));
    });
  }
  // --- JSON-RPC notification registration ---
  onNotification(method, handler) {
    this.notifications.set(method, handler);
  }
  // --- Message routing ---
  resolveMessage(raw) {
    let msg;
    try {
      msg = JSON.parse(raw);
    } catch {
      return;
    }
    if (msg.id !== void 0) {
      const pending = this.pending.get(msg.id);
      if (pending) {
        this.pending.delete(msg.id);
        if (msg.error) {
          pending.reject(msg.error);
        } else {
          pending.resolve(msg.result);
        }
      }
    } else if (msg.method) {
      const handler = this.notifications.get(msg.method);
      if (handler)
        handler(msg.params);
    }
  }
};

// src/WaveLinkController.ts
var WaveLinkController = class extends TypedEventEmitter {
  // Internal state
  inputDevices = [];
  outputDevices = [];
  mainOutput = { outputDeviceId: "" };
  channels = [];
  mixes = [];
  focusedApp = { id: "", name: "", channel: { id: "" } };
  levelMeters = {
    inputDevices: [],
    outputDevices: [],
    channels: [],
    mixes: []
  };
  _isConnected = false;
  client;
  constructor() {
    super();
    this.client = new WaveLinkClient(this);
  }
  // --- Lifecycle ---
  connect() {
    this.client.connect();
  }
  disconnect() {
    this.client.disconnect();
  }
  isConnected() {
    return this._isConnected;
  }
  // --- State accessors ---
  getInputDevices() {
    return this.inputDevices;
  }
  /** Returns only physical Wave microphone devices */
  getWaveInputDevices() {
    return this.inputDevices.filter((d) => d.isWaveDevice);
  }
  getInputDeviceById(id) {
    return this.inputDevices.find((d) => d.id === id);
  }
  getOutputDevices() {
    return this.outputDevices;
  }
  /** Returns only physical Wave output devices */
  getWaveOutputDevices() {
    return this.outputDevices.filter((d) => d.isWaveDevice);
  }
  getOutputDeviceById(id) {
    return this.outputDevices.find((d) => d.id === id);
  }
  getMainOutput() {
    return this.mainOutput;
  }
  getChannels() {
    return this.channels;
  }
  getChannelById(id) {
    return this.channels.find((c) => c.id === id);
  }
  getMixes() {
    return this.mixes;
  }
  getMixById(id) {
    return this.mixes.find((m) => m.id === id);
  }
  getLevelMeters() {
    return this.levelMeters;
  }
  getFocusedApp() {
    return this.focusedApp;
  }
  // --- Mutation methods ---
  setInputDevice(payload) {
    this.client.send("setInputDevice", payload).catch(() => {
    });
  }
  setOutputDevice(payload) {
    this.client.send("setOutputDevice", payload).catch(() => {
    });
  }
  setChannel(payload) {
    this.client.send("setChannel", payload).catch(() => {
    });
  }
  setMix(payload) {
    this.client.send("setMix", payload).catch(() => {
    });
  }
  addToChannel(appId, channelId) {
    this.client.send("addToChannel", { appId, channelId }).catch(() => {
    });
  }
  // --- Subscriptions ---
  subscribeLevelMeter(type, id, subId) {
    const payload = {
      levelMeterChanged: { isEnabled: true, type, id, ...subId ? { subId } : {} }
    };
    this.client.send("setSubscription", payload).catch(() => {
    });
  }
  unsubscribeLevelMeter(type, id, subId) {
    const payload = {
      levelMeterChanged: { isEnabled: false, type, id, ...subId ? { subId } : {} }
    };
    this.client.send("setSubscription", payload).catch(() => {
    });
  }
  subscribeFocusedApp() {
    this.client.send("setSubscription", {
      focusedAppChanged: { isEnabled: true }
    }).catch(() => {
    });
  }
  unsubscribeFocusedApp() {
    this.client.send("setSubscription", {
      focusedAppChanged: { isEnabled: false }
    }).catch(() => {
    });
  }
  // --- Internal state update methods (called by WaveLinkClient) ---
  /** @internal */
  _setConnected(val) {
    this._isConnected = val;
  }
  /** @internal */
  _setInputDevices(v) {
    this.inputDevices = v;
  }
  /** @internal */
  _setOutputDevices(v) {
    this.outputDevices = v;
  }
  /** @internal */
  _setMainOutput(v) {
    this.mainOutput = v;
  }
  /** @internal */
  _setChannels(v) {
    this.channels = v;
  }
  /** @internal */
  _setMixes(v) {
    this.mixes = v;
  }
  /** @internal */
  _setFocusedApp(v) {
    this.focusedApp = v;
  }
  /** @internal */
  _getLevelMeters() {
    return this.levelMeters;
  }
};
var WaveLinkClient = class extends BaseWaveLinkController {
  constructor(controller) {
    super();
    this.controller = controller;
  }
  async onConnect() {
    let appInfo;
    try {
      appInfo = await this.send(
        "getApplicationInfo"
      );
    } catch {
      this.ignorePortAndReconnect();
      return;
    }
    if (appInfo.appID !== "EWL" || appInfo.interfaceRevision < 1) {
      this.ignorePortAndReconnect();
      return;
    }
    const [inputRes, outputRes, channelRes, mixRes] = await Promise.all([
      this.send("getInputDevices"),
      this.send("getOutputDevices"),
      this.send("getChannels"),
      this.send("getMixes")
    ]);
    this.controller._setInputDevices(inputRes.inputDevices);
    this.controller._setOutputDevices(outputRes.outputDevices);
    this.controller._setMainOutput(outputRes.mainOutput);
    this.controller._setChannels(channelRes.channels);
    this.controller._setMixes(mixRes.mixes);
    this.controller._setConnected(true);
    this.registerNotifications();
    this.controller.emit("ready");
  }
  onDisconnect() {
    const wasConnected = this.controller.isConnected();
    this.controller._setConnected(false);
    if (wasConnected) {
      this.controller.emit("disconnected");
    }
  }
  registerNotifications() {
    this.onNotification("inputDevicesChanged", (params) => {
      this.controller._setInputDevices(params.inputDevices);
      this.controller.emit("inputDevicesChanged");
    });
    this.onNotification("inputDeviceChanged", (params) => {
      const devices = this.controller.getInputDevices();
      const target = devices.find((d) => d.id === params.id);
      if (target) {
        deepMergeDevice(target, params);
        this.controller.emit("inputDeviceChanged", params.id);
      }
    });
    this.onNotification("outputDevicesChanged", (params) => {
      const prevMain = JSON.stringify(this.controller.getMainOutput());
      if (params.mainOutput) {
        this.controller._setMainOutput(params.mainOutput);
        if (prevMain !== JSON.stringify(params.mainOutput)) {
          this.controller.emit("mainOutputDeviceChanged", params.mainOutput);
        }
      }
      if (params.outputDevices) {
        this.controller._setOutputDevices(params.outputDevices);
        this.controller.emit("outputDevicesChanged");
      }
    });
    this.onNotification("outputDeviceChanged", (params) => {
      const devices = this.controller.getOutputDevices();
      const target = devices.find((d) => d.id === params.id);
      if (target) {
        deepMergeDevice(target, params);
        this.controller.emit("outputDeviceChanged", params.id);
      }
    });
    this.onNotification("channelsChanged", (params) => {
      this.controller._setChannels(params.channels);
      this.controller.emit("channelsChanged");
    });
    this.onNotification("channelChanged", (params) => {
      const channels = this.controller.getChannels();
      const target = channels.find((c) => c.id === params.id);
      if (target) {
        const changedProps = Object.keys(params).filter(
          (k) => k !== "id" && k !== "type"
        );
        deepMergeChannel(target, params);
        this.controller.emit(
          "channelChanged",
          params.id,
          target.mixes.map((m) => m.id),
          changedProps
        );
      }
    });
    this.onNotification("mixesChanged", (params) => {
      this.controller._setMixes(params.mixes);
      this.controller.emit("mixesChanged");
    });
    this.onNotification("mixChanged", (params) => {
      const mixes = this.controller.getMixes();
      const target = mixes.find((m) => m.id === params.id);
      if (target) {
        Object.assign(target, params);
        this.controller.emit("mixChanged", params.id);
      }
    });
    this.onNotification("levelMeterChanged", (params) => {
      const lm = this.controller._getLevelMeters();
      upsertLevelMeters(lm.inputDevices, params.inputDevices ?? []);
      upsertLevelMeters(lm.outputDevices, params.outputDevices ?? []);
      upsertLevelMeters(lm.channels, params.channels ?? []);
      upsertLevelMeters(lm.mixes, params.mixes ?? []);
      this.controller.emit("levelMeterChanged");
    });
    this.onNotification("focusedAppChanged", (params) => {
      const app = {
        id: params.id,
        name: params.name === "" ? "--" : params.name,
        channel: params.channel ?? { id: "" }
      };
      this.controller._setFocusedApp(app);
      this.controller.emit("focusedAppChanged", app);
    });
  }
};
function deepMergeDevice(target, patch) {
  if (patch.inputs) {
    for (const inputPatch of patch.inputs) {
      const existing = target.inputs?.find((i) => i.id === inputPatch.id);
      if (existing)
        Object.assign(existing, inputPatch);
    }
  }
  if (patch.outputs) {
    for (const outputPatch of patch.outputs) {
      const existing = target.outputs?.find((o) => o.id === outputPatch.id);
      if (existing)
        Object.assign(existing, outputPatch);
    }
  }
  for (const key of Object.keys(patch)) {
    if (key !== "inputs" && key !== "outputs" && key !== "id") {
      target[key] = patch[key];
    }
  }
}
function deepMergeChannel(target, patch) {
  if (patch.mixes) {
    target.mixes = target.mixes.filter(
      (m) => patch.mixes.some((pm) => pm.id === m.id)
    );
    for (const mixPatch of patch.mixes) {
      const existing = target.mixes.find((m) => m.id === mixPatch.id);
      if (existing) {
        Object.assign(existing, mixPatch);
      } else {
        target.mixes.push({ ...mixPatch });
      }
    }
  }
  if (patch.effects) {
    for (const ep of patch.effects) {
      const existing = target.effects?.find((e) => e.id === ep.id);
      if (existing)
        Object.assign(existing, ep);
    }
  }
  for (const key of Object.keys(patch)) {
    if (!["mixes", "effects", "id", "type"].includes(key)) {
      target[key] = patch[key];
    }
  }
}
function upsertLevelMeters(arr, incoming) {
  for (const item of incoming) {
    const idx = arr.findIndex(
      (x) => x.id === item.id && x.subId === item.subId
    );
    if (idx >= 0) {
      arr[idx] = item;
    } else {
      arr.push(item);
    }
  }
}
export {
  WaveLinkController
};
